File: Serene/Serene.Script/Common/Helper/ReportHelper.cs

namespace Serene.Common
{
    using Serenity;
    using System;
    using System.Runtime.CompilerServices;

    [Imported]
    public static class ReportHelper
    {
        public static ToolButton CreateToolButton(ReportButtonOptions options)
        {
            return null;
        }
    }

    [Imported, Serializable]
    public class ReportButtonOptions
    {
        public bool Download { get; set; }
        public string Title { get; set; }
        public string CssClass { get; set; }
        public string Icon { get; set; }
        public string ReportKey { get; set; }
        public string Extension { get; set; }
        public Func<object> GetParams { get; set; }
        public string Target { get; set; }
    }
}

File: Serene/Serene.Web/Modules/Common/Helpers/ReportHelper.ts
namespace Serene.Common {

    export interface ReportButtonOptions {
        title?: string;
        cssClass?: string;
        icon?: string;
        download?: boolean;
        reportKey: string;
        extension?: string;
        getParams?: () => any;
        target?: string;
    }

    export namespace ReportHelper {

        export function createToolButton(options: ReportButtonOptions): Serenity.ToolButton {
            return {
                title: Q.coalesce(options.title, 'Report'),
                cssClass: Q.coalesce(options.cssClass, 'print-button'),
                icon: options.icon,
                onClick: () => {
                    Q.postToUrl({
                        url: '~/Report/' + (options.download ? 'Download' : 'Render'),
                        params: {
                            key: options.reportKey,
                            ext: Q.coalesce(options.extension, 'pdf'),
                            opt: (options.getParams == null ? '' : $.toJSON(options.getParams()))
                        },
                        target: Q.coalesce(options.target, '_blank')
                    });
                }
            };
        }
    }
}